### DISCUSSION TITLE
It reads all the records in the table, although I indicate the key to bring only 1 record.

### DISCUSSION THREAD

**Discussion by patriciomh1965** [serenity-is/Serenity#5396] [Help] [answered]
Hello Community

I have a reading problem in the XYZGrid. I am reading a table to obtain 1 record, the data obtained is transferred to a form. But it brings me all the records in the table, which is why it takes a long time to open the form.

I would appreciate the help you can offer me. Atte.-

![grilla](https://user-images.githubusercontent.com/42218266/105706740-b19dc180-5ef0-11eb-8672-bce7af56fa1d.png)

![cuadro](https://user-images.githubusercontent.com/42218266/105708112-b794a200-5ef2-11eb-8d34-c66c26adecf8.png)

Next I show you the image and the classes:

**_XYZGrid :_**

namespace SegurosMAPC.BaseDatos {

    @Serenity.Decorators.registerClass()
    export class MonitorSegurosMaestrosGrid extends Serenity.EntityGrid<MonitorSegurosMaestrosRow, any> {
        protected getColumnsKey() { return 'BaseDatos.MonitorSegurosMaestros'; }
        protected getDialogType() { return MonitorSegurosMaestrosDialog; }
        protected getIdProperty() { return MonitorSegurosMaestrosRow.idProperty; }
        protected getInsertPermission() { return MonitorSegurosMaestrosRow.insertPermission; }
        protected getLocalTextPrefix() { return MonitorSegurosMaestrosRow.localTextPrefix; }
        protected getService() { return MonitorSegurosMaestrosService.baseUrl; }

        constructor(container: JQuery) {
            super(container);
        }
        
        getButtons() {
            var buttons = super.getButtons();

            buttons.splice(Q.indexOf(buttons, x => x.cssClass == "add-button"), 1); // QUITAR BOTON NUEVO

            buttons.push(SegurosMAPC.Common.ExcelExportHelper.createToolButton({
                title: "Exportar",
                grid: this,
                onViewSubmit: () => this.onViewSubmit(),
                service: 'BaseDatos/MonitorSegurosMaestros/ListExcel',
                separator: true
            }));          

            buttons.push({
                title: '',
                icon: 'fa fa-filter text-blue',
                onClick: () => {
                    this.quickFiltersDiv.slideToggle();
                }
            });

            return buttons;
        }

        //  APROBACION POR LINEAS
        protected getColumns() {
            var columns = super.getColumns();           

            columns.splice(0, 0,
                {
                    field: 'DETALLE DEL SEGURO',
                    name: '',
                    format: ctx => '<a class="inline-action detalle" title="Detalle del Seguro">' +
                        '<i class="fa fa-credit-card text-blue"></i></a>',
                    width: 24,
                    minWidth: 24,
                    maxWidth: 24
                });
            return columns;
        }

        protected onClick(e: JQueryEventObject, row: number, cell: number)
        {
            super.onClick(e, row, cell);
            if (e.isDefaultPrevented())
                return;

            var item = this.itemAt(row);
            var ramo = item.MseRamo;
            var target = $(e.target);
            if (target.parent().hasClass('inline-action'))
                target = target.parent();

            if (target.hasClass('inline-action'))
            {
                e.preventDefault();

                if (target.hasClass('detalle'))
                {
                    var $this = this;

                    // Riesgo
                    if (ramo == 1) {

**_HERE IS THE READING TO THE TABLE :_**

                       Q.reloadLookup(BaseDatos.TaMaeUbicacionesRiesgoRow.lookupKey);
                        var datariesgo = TaMaeUbicacionesRiesgoRow.getLookup().itemById[item.MseNumeroSeguro];
                        if (datariesgo == null) { Q.notifyError("Seguro no tiene la materia de Riesgo"); return; }
**_ END READING:_**

                        var dlg = new MonitorSegurosRiesgosDialog();
                        this.initDialog(dlg);
                        dlg.loadEntityAndOpenDialog(<MonitorSegurosRiesgosRow>{
                            // seguro
                            MseNroContrato: item.MseNroContrato,
                            MseCodMaterial: item.MseCodMaterial,
                            MseTipoSeguro: item.MseTipoSeguro,
                            MseNumeroSeguro: item.MseNumeroSeguro,
                            MseDeducible: item.MseDeducible,
                            MsePrimaBruta: item.MsePrimaBruta,
                            MsePrimaAfecta: item.MsePrimaAfecta,
                            MsePrimaExenta: item.MsePrimaExenta,
                            MseMesesVigencia: item.MseMesesVigencia,
                            MseCantidadCuotas: item.MseCantidadCuotas,
                            MseDiasVigencia: item.MseDiasVigencia,
                            MseFechaInicioVigencia: item.MseFechaInicioVigencia,
                            MseFechaFinVigencia: item.MseFechaFinVigencia,
                            MseRubro: item.MseRubro,
                            MseDesRubro: item.MseDesRubro,
                            MseSubrubro: item.MseSubrubro,
                            MseNumeroMaterial: item.MseNumeroMaterial,
                            MseGrupoMaterial: item.MseGrupoMaterial,
                            MseGiroEmpresa: item.MseGiroEmpresa,
                            MseSumaAsegurada: item.MseSumaAsegurada,
                            MseAccesorio: item.MseAccesorio,
                            MseNumeroUnidades: item.MseNumeroUnidades,
                            MseTipoAseguramiento: item.MseTipoAseguramiento,
                            MseDesMateriaAsegurada: item.MseDesMateriaAsegurada,
                            MseFormaPago: item.MseFormaPago,
                            MseFechaMeraTenencia: item.MseFechaMeraTenencia,
                            MseComisionesEspeciales: item.MseComisionesEspeciales,
                            MseNotasAclaratorias: item.MseNotasAclaratorias,
                            MseEstadoSeguro: item.MseEstadoSeguro,
                            MseEstadoTransaccion: item.MseEstadoTransaccion,
                            MseFechaTransaccion: item.MseFechaTransaccion,
                            MseCodUltimaTxInyectada: item.MseCodUltimaTxInyectada,
                            MseNumeroVigencia: item.MseNumeroVigencia,
                            MpeRut: item.MpeRut,
                            MseTipoProducto: item.MseTipoProducto,
                            MseCobertura: item.MseCobertura,
                            MseNumeroProceso: item.MseNumeroProceso,
                            MseNumeroLinea: item.MseNumeroLinea,
                            MseFechaProceso: item.MseFechaProceso,
                            // detalle incendio
                            Calle: datariesgo.Calle,
                            Piso: datariesgo.Piso,
                            Numero: datariesgo.Numero,
                            CodigoRegionIsol: datariesgo.CodigoRegionIsol,
                            CodigoCiudadIsol: datariesgo.CodigoCiudadIsol,
                            CodigoComunaIsol: datariesgo.CodigoComunaIsol,
                            Departamento: datariesgo.Departamento,
                            CodigoPostal: datariesgo.CodigoPostal
                        });
                    }
                    // Vehiculo
                    if (ramo == 2)
                    {
                        Q.reloadLookup(BaseDatos.TaMaeVehiculoRow.lookupKey);
                        var datavehiculo = TaMaeVehiculoRow.getLookup().itemById[item.MseNumeroSeguro];
                        if (datavehiculo == null) { Q.notifyError("Seguro no tiene la materia de Vehiculo"); return; }  // 
                        var dlg2 = new MonitorSegurosVehiculoDialog();
                        this.initDialog(dlg2);
                        dlg2.loadEntityAndOpenDialog(<MonitorSegurosVehiculoRow>{
                            // seguro
                            MseNroContrato: item.MseNroContrato,
                            MseCodMaterial: item.MseCodMaterial,
                            MseTipoSeguro: item.MseTipoSeguro,
                            MseNumeroSeguro: item.MseNumeroSeguro,
                            MseDeducible: item.MseDeducible,
                            MsePrimaBruta: item.MsePrimaBruta,
                            MsePrimaAfecta: item.MsePrimaAfecta,
                            MsePrimaExenta: item.MsePrimaExenta,
                            MseMesesVigencia: item.MseMesesVigencia,
                            MseCantidadCuotas: item.MseCantidadCuotas,
                            MseDiasVigencia: item.MseDiasVigencia,
                            MseFechaInicioVigencia: item.MseFechaInicioVigencia,
                            MseFechaFinVigencia: item.MseFechaFinVigencia,
                            MseRubro: item.MseRubro,
                            MseDesRubro: item.MseDesRubro,
                            MseSubrubro: item.MseSubrubro,
                            MseNumeroMaterial: item.MseNumeroMaterial,
                            MseGrupoMaterial: item.MseGrupoMaterial,
                            MseGiroEmpresa: item.MseGiroEmpresa,
                            MseSumaAsegurada: item.MseSumaAsegurada,
                            MseAccesorio: item.MseAccesorio,
                            MseNumeroUnidades: item.MseNumeroUnidades,
                            MseTipoAseguramiento: item.MseTipoAseguramiento,
                            MseDesMateriaAsegurada: item.MseDesMateriaAsegurada,
                            MseFormaPago: item.MseFormaPago,
                            MseFechaMeraTenencia: item.MseFechaMeraTenencia,
                            MseComisionesEspeciales: item.MseComisionesEspeciales,
                            MseNotasAclaratorias: item.MseNotasAclaratorias,
                            MseEstadoSeguro: item.MseEstadoSeguro,
                            MseEstadoTransaccion: item.MseEstadoTransaccion,
                            MseFechaTransaccion: item.MseFechaTransaccion,
                            MseCodUltimaTxInyectada: item.MseCodUltimaTxInyectada,
                            MseNumeroVigencia: item.MseNumeroVigencia,
                            MpeRut: item.MpeRut,
                            MseTipoProducto: item.MseTipoProducto,
                            MseCobertura: item.MseCobertura,
                            MseNumeroProceso: item.MseNumeroProceso,
                            MseNumeroLinea: item.MseNumeroLinea,
                            MseFechaProceso: item.MseFechaProceso,
                            // detalle vehiculo
                            IdTipoVehiculo: datavehiculo.IdTipoVehiculo,
                            Chasis: datavehiculo.Chasis,
                            Motor: datavehiculo.Motor,
                            ColorIsol: datavehiculo.ColorIsol,
                            RutEmisor: datavehiculo.RutEmisor,
                            NombreFacturacion: datavehiculo.NombreFacturacion,
                            NroFactura: datavehiculo.NroFactura,
                            Patente: datavehiculo.Patente,
                            Es0km: datavehiculo.Es0km,
                            AnoFabricacion: datavehiculo.AnoFabricacion,
                            FechaEmision: datavehiculo.FechaEmision,
                            CodigoUso: datavehiculo.CodigoUso,
                            NumeroAsientoPasejero: datavehiculo.NumeroAsientoPasejero,
                            MarcaEle: datavehiculo.MarcaEle,
                            ModeloEle: datavehiculo.ModeloEle,                            
                            ColorEle: datavehiculo.ColorEle,      
                        });
                    }

                    // Objeto Valioso
                    if (ramo == 3) {                       
                        var dlg3 = new MonitorSegurosObjetoDialog();
                        this.initDialog(dlg3);
                        dlg3.loadEntityAndOpenDialog(<MonitorSegurosObjetoRow>{ 
                            // seguro
                            MseNroContrato: item.MseNroContrato,
                            MseCodMaterial: item.MseCodMaterial,
                            MseTipoSeguro: item.MseTipoSeguro,
                            MseNumeroSeguro: item.MseNumeroSeguro,
                            MseDeducible: item.MseDeducible,
                            MsePrimaBruta: item.MsePrimaBruta,
                            MsePrimaAfecta: item.MsePrimaAfecta,
                            MsePrimaExenta: item.MsePrimaExenta,
                            MseMesesVigencia: item.MseMesesVigencia,
                            MseCantidadCuotas: item.MseCantidadCuotas,
                            MseDiasVigencia: item.MseDiasVigencia,
                            MseFechaInicioVigencia: item.MseFechaInicioVigencia,
                            MseFechaFinVigencia: item.MseFechaFinVigencia,
                            MseRubro: item.MseRubro,
                            MseDesRubro: item.MseDesRubro,
                            MseSubrubro: item.MseSubrubro,
                            MseNumeroMaterial: item.MseNumeroMaterial,
                            MseGrupoMaterial: item.MseGrupoMaterial,
                            MseGiroEmpresa: item.MseGiroEmpresa,
                            MseSumaAsegurada: item.MseSumaAsegurada,
                            MseAccesorio: item.MseAccesorio,
                            MseNumeroUnidades: item.MseNumeroUnidades,
                            MseTipoAseguramiento: item.MseTipoAseguramiento,
                            MseDesMateriaAsegurada: item.MseDesMateriaAsegurada,
                            MseFormaPago: item.MseFormaPago,
                            MseFechaMeraTenencia: item.MseFechaMeraTenencia,
                            MseComisionesEspeciales: item.MseComisionesEspeciales,
                            MseNotasAclaratorias: item.MseNotasAclaratorias,
                            MseEstadoSeguro: item.MseEstadoSeguro,
                            MseEstadoTransaccion: item.MseEstadoTransaccion,
                            MseFechaTransaccion: item.MseFechaTransaccion,
                            MseCodUltimaTxInyectada: item.MseCodUltimaTxInyectada,
                            MseNumeroVigencia: item.MseNumeroVigencia,
                            MpeRut: item.MpeRut,
                            MseTipoProducto: item.MseTipoProducto,
                            MseCobertura: item.MseCobertura,
                            MseNumeroProceso: item.MseNumeroProceso,
                            MseNumeroLinea: item.MseNumeroLinea,
                            MseFechaProceso: item.MseFechaProceso,
                        });
                    }
                }
            }
        }
    }
}

**_XYZRow_**

namespace SegurosMAPC.BaseDatos.Entities
{
    using Serenity;
    using Serenity.ComponentModel;
    using Serenity.Data;
    using Serenity.Data.Mapping;
    using System;
    using System.ComponentModel;
    using System.IO;

    [ConnectionKey("BaseDatos"), Module("BaseDatos"), TableName("[dbo].[ta_mae_ubicaciones_riesgo]")]
    [DisplayName("Maestro Ubicaciones Riesgo"), InstanceName("Maestro Ubicaciones Riesgo")]
    [ReadPermission("Administration:General")]
    [ModifyPermission("Administration:General")]

    [LookupScript(Expiration = -1)]
    public sealed class TaMaeUbicacionesRiesgoRow : Row, IIdRow, INameRow
    {

        [DisplayName("Numero Seguro"), Column("mur_numero_seguro"), PrimaryKey]
        public Int32? NumeroSeguro
        {
            get { return Fields.NumeroSeguro[this]; }
            set { Fields.NumeroSeguro[this] = value; }
        }

        [DisplayName("Id Domicilio"), Column("mur_id_domicilio"), PrimaryKey]
        public Int32? IdDomicilio
        {
            get { return Fields.IdDomicilio[this]; }
            set { Fields.IdDomicilio[this] = value; }
        }

        [DisplayName("Tipo Direccion"), Column("mur_tipo_direccion"), Size(1), LookupInclude]
        public String TipoDireccion
        {
            get { return Fields.TipoDireccion[this]; }
            set { Fields.TipoDireccion[this] = value; }
        }

        [DisplayName("Calle"), Column("mur_calle"), Size(100), LookupInclude]
        public String Calle
        {
            get { return Fields.Calle[this]; }
            set { Fields.Calle[this] = value; }
        }

        [DisplayName("Numero"), Column("mur_numero"), Size(20), LookupInclude]
        public String Numero
        {
            get { return Fields.Numero[this]; }
            set { Fields.Numero[this] = value; }
        }

        [DisplayName("Piso"), Column("mur_piso"), Size(5), LookupInclude]
        public String Piso
        {
            get { return Fields.Piso[this]; }
            set { Fields.Piso[this] = value; }
        }

        [DisplayName("Departamento"), Column("mur_departamento"), Size(20), LookupInclude]
        public String Departamento
        {
            get { return Fields.Departamento[this]; }
            set { Fields.Departamento[this] = value; }
        }

        [DisplayName("Codigo Postal"), Column("mur_codigo_postal"), Size(15), LookupInclude]
        public String CodigoPostal
        {
            get { return Fields.CodigoPostal[this]; }
            set { Fields.CodigoPostal[this] = value; }
        }

        [DisplayName("Codigo Comuna Isol"), Column("mur_codigo_comuna_isol"), LookupInclude]
        public Int32? CodigoComunaIsol
        {
            get { return Fields.CodigoComunaIsol[this]; }
            set { Fields.CodigoComunaIsol[this] = value; }
        }

        [DisplayName("Codigo Region Isol"), Column("mur_codigo_region_isol"), LookupInclude]
        public Int32? CodigoRegionIsol
        {
            get { return Fields.CodigoRegionIsol[this]; }
            set { Fields.CodigoRegionIsol[this] = value; }
        }

        [DisplayName("Codigo Ciudad Isol"), Column("mur_codigo_ciudad_isol"), LookupInclude]
        public Int32? CodigoCiudadIsol
        {
            get { return Fields.CodigoCiudadIsol[this]; }
            set { Fields.CodigoCiudadIsol[this] = value; }
        }

        [DisplayName("Numero Material"), Column("mur_numero_material"), LookupInclude]
        public Int32? NumeroMaterial
        {
            get { return Fields.NumeroMaterial[this]; }
            set { Fields.NumeroMaterial[this] = value; }
        }

        [DisplayName("Grupo Material"), Column("mur_grupo_material"), LookupInclude]
        public Int32? GrupoMaterial
        {
            get { return Fields.GrupoMaterial[this]; }
            set { Fields.GrupoMaterial[this] = value; }
        }

        [DisplayName("Ciudad Riesgo"), Column("mur_ciudad_riesgo"), Size(20), LookupInclude]
        public String CiudadRiesgo
        {
            get { return Fields.CiudadRiesgo[this]; }
            set { Fields.CiudadRiesgo[this] = value; }
        }

        [DisplayName("Id"), Column("mur_id"), Identity]
        public Int32? Id
        {
            get { return Fields.Id[this]; }
            set { Fields.Id[this] = value; }
        }

        IIdField IIdRow.IdField
        {
            get { return Fields.NumeroSeguro; }
        }

        StringField INameRow.NameField
        {
            get { return Fields.TipoDireccion; }
        }

        public static readonly RowFields Fields = new RowFields().Init();

        public TaMaeUbicacionesRiesgoRow()
            : base(Fields)
        {
        }

        public class RowFields : RowFieldsBase
        {
            public Int32Field NumeroSeguro;
            public Int32Field IdDomicilio;
            public StringField TipoDireccion;
            public StringField Calle;
            public StringField Numero;
            public StringField Piso;
            public StringField Departamento;
            public StringField CodigoPostal;
            public Int32Field CodigoComunaIsol;
            public Int32Field CodigoRegionIsol;
            public Int32Field CodigoCiudadIsol;
            public Int32Field NumeroMaterial;
            public Int32Field GrupoMaterial;
            public StringField CiudadRiesgo;
            public Int32Field Id;
        }
    }
}

**Comment by wezmag** [comment]
When you call `getLookup()`, it will load all records from either database or cache. And because you use `Q.reloadLookup`, so it loads from database.

To fix the issue, you might want to use `TaMaeUbicacionesRiesgoService.Retrieve` to obtain only 1 record.

**Reply by edwardch** [reply]
Correct, using `TaMaeUbicacionesRiesgoService.Retrieve` will do a trip the database and fetch only this one record.

```ts
var d = BaseDatos.TaMaeUbicacionesRiesgoService.Retrieve({
    EntityId: item.MseNumeroSeguro // or whatever your Id field is for this type
}, response => {
    this.form.SomeField.value = response.Entity.SomeField;
    this.form.SomeForeignKeyId.value = Q.toId(response.Entity.SomeForeignKeyId);
    // you can do anything in the response, really
});
```

Good luck!

**Comment by patriciomh1965** [comment, marked as answer]
Thanks for your help, I will implement it. Atte.-

**Reply by edwardch** [reply]
Please can you mark this as answered to help other community members in future with a similar question?
